### ISSUE TITLE
Run local linter / formatter

### ISSUE THREAD

**Issue by ZelCloud** [raxod502/apheleia#88]
Some javascript projects and repos have prettier used to format the code, then they'll use eslint to do some customized format changes on top of the formatted prettier code.

Example: package.json setup
```json
"prettier": "prettier-config-cust",
"eslintConfig": {
	"extends": [
	   	"eslint-config-cust"
	]
},
```

Then they'll format the code using the following command

```bash
npm eslint {{path_to_file}} --fix
```

The above will run prettier first then eslint to auto format some other rules. Also this command depends on the directory path since it will reference the package.json above for the formatting rules.

So my question, is there a way to setup running the local version of a formatter and not a global instance?

```elisp
;; Global instance of prettier is ran with a setup like so

(use-package apheleia
  :ensure t
  :defer t
  :config

  (setf (alist-get 'prettier apheleia-formatters)
      '("prettier"))

  (push '(prettier . ("prettier"
                      file
		      ))
		apheleia-formatters)

  (setf (alist-get 'web-mode apheleia-mode-alist)
		'(prettier))
  )
```

**Comment by ZelCloud** [comment]
Duplicate of #30 so closing this issue.

**Comment by raxod502** [comment]
Apheleia is configured to do this by default, using the `npx` symbol documented in `apheleia-formatters`. Does your local configuration take effect if you remove your custom configuration?

**Comment by ZelCloud** [comment]
I'm sorry I probably didn't explain the issue / request well enough, or I might be misunderstanding something. 

The default apheleia prettier setup uses npx which uses the global prettier instance formatter which is mostly fine when prettier is the only formatter needed to be used. I can also confirm running prettier this way does use a local prettier.config.js file if it exists in the repo.

The issue is this doesn't use any of the eslint rules (lint / format changes) on top. In order to do this properly (run prettier first then eslint) locally we need to run the following command

```bash
npm run eslint {{file}} --fix
```

When you have both the prettier config and eslint config extending prettier, eslint --fix will actually run prettier first to get the initial formatting then run itself with the --fix flag (lint format according to eslint rules). It's also important to use the local prettier and eslint copies since different versions have slightly different behaviors when it comes to formatting and depending on the frontend framework being used, so global prettier + eslint combo isn't a good option.

So in the default apheleia case we have 

1. prettier (global - kicked off by the command - npx prettier)

while I need

1. prettier (local - kicked off by the command - using npm run eslint --fix)
2. eslint --fix (local)

So in this case wouldn't I need a custom config something like

```elisp
(defun javascript-get-root-directory ()
 	(locate-dominating-file (or buffer-file-name default-directory) "package.json")
 	)

;; not exactly sure how to tell formatter to run within a specific directory - in order to get the correct npm run context
  (push '(prettier . ("npm run eslint" ;; run within the root directory of the project somehow - (javascript-get-root-directory)
                      file
                      "--fix"
		      ))
		apheleia-formatters)
```

**Comment by raxod502** [comment]
I see - but presumably this `npm run eslint --fix` command would fail unless you happened to be within a project directory whose `package.json` defined the appropriate scripts?

Anyway, it's a valid use case. Currently, we don't have a way of explicitly determining the working directory using arbitrary Elisp, but we should. Retitling this as a feature request for that.

I think the easiest workaround until that's implemented would be to create a shell script on your `$PATH` that would change working directory as necessary, before then invoking the provided command. Something like `npm-run-from-root`. Then you could reference that in your formatter definition.

**Comment by ZelCloud** [comment]
> I see - but presumably this npm run eslint --fix command would fail unless you happened to be within a project directory whose package.json defined the appropriate scripts?

Yes, you're correct in your assumption, if we run the command outside of the root directory of the project the command will fail.

> Anyway, it's a valid use case. Currently, we don't have a way of explicitly determining the working directory using arbitrary Elisp, but we should. Retitling this as a feature request for that.

❤️- Ya it's easy to determine a lot of project root directory's from our side (user side) we just need a way to forward it to apheleia (package side).

```elisp
;; examples
;; js - package.json projects root
(defun javascript-get-root-directory ()
 	(locate-dominating-file (or buffer-file-name default-directory) "package.json")
 	)

;; rustfmt - rust project root
(defun rust-get-root-directory ()
 	(locate-dominating-file (or buffer-file-name default-directory) "cargo.toml")
 	)
```

**Comment by tilgovi** [comment]
I'm not sure I follow this issue. Couldn't `npx` also be used to run `eslint`? Doesn't a global ESLint also delegate to the project-local ESLint version, if it's installed locally?

**Comment by tilgovi** [comment]
Maybe I can contribute something useful to this discussion.

@ZelCloud, it sounds like you are referring to an ESLint configuration that uses `eslint-plugin-prettier`, because you say:

> prettier (local - kicked off by the command - using npm run eslint --fix)

If that is the case, there's no reason to run ESLint again, afterwards.

But this pattern is discouraged by prettier these days. The prettier project recommends using `eslint-config-prettier` that disables rules that conflict with prettier, and then running them in sequence (`prettier`, than `eslint`).

Both ESLint and Prettier discover configuration relative to the location of the input file. The working directory should not matter. Apheleia should not need to change the working directory to make these tools work.

Finding the right command to run is something where there may be room for improvement, though. The `npx` symbol currently is trying to discover the `node_modules/.bin` path based on the input file, but it's maybe possible to use the `npx` CLI tool to do this instead. There's probably a similar thing to be done with `yarn`, and supporting Yarn P'n'P is another step further, even, but I'm interested in seeing what I can do to help.

**Comment by tilgovi** [comment]
To repeat that more succinctly, Prettier and ESLint don't care about the working directory, but `npm run` does.

**Comment by tilgovi** [comment]
If we do want to choose a working directory that's different from the directory of the input file, we could consider using project.el's project directory. That's already customizable in ways that are pluggable. It would generally result in a sane default for VCS projects, and folks using something like projectile would get support for a bunch of languages out of the box.

**Comment by jesse-c** [comment]
> If we do want to choose a working directory that's different from the directory of the input file, we could consider using project.el's project directory. That's already customizable in ways that are pluggable. It would generally result in a sane default for VCS projects, and folks using something like projectile would get support for a bunch of languages out of the box.

This would be fantastic. For example, `mix format` looks for a `.formatter.exs` file. I have the issue now where apheleia's output is different to the project's configured formatting. You can use a `--dot-formatter $file` arg for `mix format`. I'm looking at for now, changing `mix-format` to include a check if the there's a project root set and then using that.

**Comment by mkeen** [comment]
project.el hooks seem like they'd be apt here. i've got some trash ones i have cobbled together myself, so not sharing (they are truly bad) but to me, project.el is a good way to get current project root working directory across any language/framework.
